File: src/models/applicant.ts


export enum AvailabilityEnum {
    IMMEDIATE = "immediate",
    WITHIN_3_MONTHS = "within_3_months",
    WITHIN_6_MONTHS = "within_6_months",
    WITHIN_1_YEAR = "within_1_year",
    IF_I_FIND_A_GOOD_COMPANY = "if_I_find_a_good_company",
    NOT_THINKING_ABOUT_IT_NOW = "not_thinking_about_it_now"
}

export enum GenderEnum {
    MALE = "male",
    FEMALE = "female",
    OTHER = "other"
}

export enum ManagementExperienceYearEnum {
    YEARS_0 = "years_0",
    YEARS_1 = "years_1",
    YEARS_2 = "years_2",
    YEARS_3 = "years_3",
    YEARS_4 = "years_4",
    YEARS_5 = "years_5",
    YEARS_6 = "years_6",
    YEARS_7 = "years_7",
    YEARS_8 = "years_8",
    YEARS_9 = "years_9",
    YEARS_10 = "years_10",
    YEARS_10_MORE = "years_10_more"
}

export type ApplicantModel = {
      id: number;
      created_at: Date;
      updated_at: Date;
      full_name: string;
      furigana: string;
      phone_number: string;
      birthdate: Date;
      location_id: number;
      address: string;
      income_range_id: number;
      availability: `${AvailabilityEnum}`;
      resume_file: any;
      description: string;
      management_experience_id: number;
      encrypted_password: string;
      email: string;
      reset_password_token: string;
      reset_password_sent_at: Date;
      remember_created_at: Date;
      current_sign_in_at: Date;
      last_sign_in_at: Date;
      current_sign_in_ip: string;
      last_sign_in_ip: string;
      sign_in_count: number;
      password: string;
      password_confirmation: string;
      locked_at: Date;
      failed_attempts: number;
      unlock_token: string;
      gender: `${GenderEnum}`;
      confirmation_token: string;
      unconfirmed_email: string;
      confirmed_at: Date;
      confirmation_sent_at: Date;
      is_onboarded: boolean;
      notes: string;
      current_company_name: string;
      current_company_department: string;
      current_company_position: string;
      confirmation_email_resent_count: number;
      number_of_companies: number;
      is_suspended: boolean;
      desired_info: string;
      management_experience_years: `${ManagementExperienceYearEnum}`;
  
}

File: app/models/applicant.rb
class Applicant < ApplicationRecord
  devise :database_authenticatable, :registerable, :rememberable, :validatable,
         :trackable, :recoverable, :lockable, :confirmable

  has_many :applicant_industries, dependent: :destroy
  has_many :applicant_schools, dependent: :destroy
  has_many :applicant_languages, dependent: :destroy
  has_many :messages, dependent: :destroy
  has_many :job_applications, dependent: :destroy
  has_many :job_experiences, dependent: :destroy
  has_many :skills, dependent: :destroy
  has_many :job_opening_favorites, dependent: :destroy
  has_many :restrictions, dependent: :destroy
  has_many :recruitment_notes, dependent: :destroy
  has_many :target_list_applicants, dependent: :destroy

  belongs_to :location
  belongs_to :income_range
  belongs_to :management_experience

  enum availability: %w[immediate within_3_months within_6_months within_1_year if_I_find_a_good_company not_thinking_about_it_now],
       _suffix: true
  enum gender: %w[male female other], _suffix: true
  enum management_experience_years: %w[years_0 years_1 years_2 years_3 years_4 years_5 years_6 years_7 years_8 years_9 years_10 years_10_more],
       _suffix: true

  has_one_attached :resume_file, dependent: :destroy

  # validations

  PASSWORD_FORMAT = /\A(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*-]).{8,}\z/
  validates :password, format: PASSWORD_FORMAT, if: -> { new_record? || password.present? }

  validates :full_name, length: { in: 0..255 }

  validates :furigana, length: { in: 0..255 }

  validates :phone_number, length: { in: 0..0 }

  validates :phone_number, format: { with: /Ad{10}z/ }

  validates :birthdate, presence: true, timeliness: { type: :date, on_or_before: Date.yesterday }

  validates :address, length: { in: 0..255 }

  validates :resume_file, content_type: ['application/pdf'], size: { less_than_or_equal_to: 100.megabytes }

  validates :description, length: { in: 0..250 }

  validates :email, presence: true, uniqueness: true

  validates :email, length: { in: 0..255 }

  validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }

  validates :notes, length: { in: 0..255 }

  validates :current_company_name, length: { in: 0..255 }

  validates :current_company_department, length: { in: 0..255 }

  validates :current_company_position, length: { in: 0..255 }

  validates :confirmation_email_resent_count,
            numericality: { greater_than_or_equal_to: 0.0, less_than_or_equal_to: 0.0 }

  validates :number_of_companies, numericality: { greater_than_or_equal_to: 0.0, less_than_or_equal_to: 99.0 }

  validates :desired_info, length: { in: 0..65_535 }

  # end for validations

  def generate_reset_password_token
    raw, enc = Devise.token_generator.generate(self.class, :reset_password_token)
    self.reset_password_token   = enc
    self.reset_password_sent_at = Time.now.utc
    save(validate: false)
    raw
  end

  class << self
    def authenticate?(email, password)
      applicant = Applicant.find_for_authentication(email: email)
      return false if applicant.blank?

      if applicant&.valid_for_authentication? { applicant.valid_password?(password) }
        applicant.reset_failed_attempts!
        return applicant
      end

      # We will show the error message in TokensController
      return applicant if applicant&.access_locked?

      false
    end

    def verify_otp?(phone_number, otp_code)
      phone_number = ::Auths::PhoneNumber.new({ phone_number: phone_number, otp_code: otp_code })
      return unless phone_number.valid?

      ::Auths::PhoneVerification.new(phone_number.formatted_phone_number).verify_otp(otp_code)
      find_by(phone_number: phone_number.formatted_phone_number)
    end
  end
end
